### PR TITLE
West compare command

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
Help text for the new command:

```
$ west help compare
usage: west compare [-h] [-a] [--exit-code] [--ignore-branches] [PROJECT ...]

Compare each project's working tree state against the results
of the most recent successful "west update" command.

This command prints "git status" output for a project if
(and only if) at least one of the following is true:

1. its checked out commit is different than the commit checked
   out by the most recent successful "west update"
2. its working tree is not clean (i.e. there are local uncommitted
   changes)
3. it has a local checked out branch

positional arguments:
  PROJECT            projects (by name or path) to operate on; defaults to active cloned projects

options:
  -h, --help         show this help message and exit
  -a, --all          include output for inactive projects
  --exit-code        exit with status code 1 if status output was printed for any project
  --ignore-branches  skip output for projects with checked out branches and clean
                     working trees if the branch is at the same commit as the
                     last "west update"

ACTIVE PROJECTS
---------------

Default output is limited to "active" projects as determined by the:

- "group-filter" manifest file section
- "manifest.group-filter" local configuration option in .west/config

To include inactive projects as well, use "--all" or give an explicit
list of projects (by name or path). See the west documentation for
more details on active projects.

Regardless of the above, output is limited to cloned projects.
```

Fixes: #642

### CODE REVIEW

**Comment by mbolivar-nordic** [comment]
Putting a DNM on this until we get the requester to confirm it's what they want, and then I'll write some test cases.

**Comment by marc-hb** [comment]
That's really nice but it seems to still suffer from git's "non-determistic" display behavior which I analyzed and documented a couple years ago in https://github.com/zephyrproject-rtos/west/issues/548#issuecomment-961624310

This git behavior makes it show `manifest-rev` sometimes and sometimes not - even when everything is on the `manifest-rev`. I was lucky enough to demonstrate this in just 5 minutes:

```
SOF$ west compare

=== status of hal_xtensa (modules/hal/xtensa):
HEAD detached at refs/heads/manifest-rev
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	foo

nothing added to commit but untracked files present (use "git add" to track)

=== status of mbedtls (modules/crypto/mbedtls):
HEAD detached at 6e1660500756
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	foo

nothing added to commit but untracked files present (use "git add" to track)
```
The output above really looks like `hal_xtensa` is on the manifest while `mbedtls` is not on the manifest.

WRONG! That's gits "non-deterministic" display deceiving you. In reality they're both on the manifest:

```
git -C modules/crypto/mbedtls/ log --oneline --decorate -n 1
6e1660500756 (HEAD, manifest-rev) Merge pull request for mbedtls-3.2.1
```


PS: `west compare` sounds like it could be extended in the future to "recursively" compare manifests, thoughts on that?

**Comment by mbolivar-nordic** [comment]
> it seems to still suffer from git's "non-deterministic" display behavior

I am personally OK with this as a limitation unless there is a way to tell git status to try harder to avoid this.

> PS: west compare sounds like it could be extended in the future to "recursively" compare two manifests, thoughts on that?

The name `compare` was not arbitrary ;)

https://github.com/nrfconnect/sdk-nrf/blob/57fade75665b2d6e1af649ff47a72374dba31a0b/scripts/west_commands/ncs_commands.py#L400

**Comment by marc-hb** [comment]
> unless there is a way to tell git status to try harder to avoid this.

I didn't find any; that's frustrating. `git log` has `--decorate` but nothing in "non-deterministic" `git status`

How about this one-liner? Works4me!
```diff
--- a/src/west/app/project.py
+++ b/src/west/app/project.py
@@ -682,6 +682,9 @@ class Compare(_ProjectCommand):
                     continue
 
                 self.banner(f'status of {project.name_and_path}:')
+                # `git status` shows `manifest-rev` "sometimes", see #643
+                # If manifest-rev is missing then we already failed earlier anyway
+                project.git('log --oneline -n 1 manifest-rev')
                 project.git('status')
                 printed_status = True
             except subprocess.CalledProcessError:
```

> https://github.com/nrfconnect/sdk-nrf/blob/57fade75665b2d6e1af649ff47a72374dba31a0b/scripts/west_commands/ncs_commands.py#L400

Way ahead of "us" upstream! I didn't look at the code much but the name in the git history seems very familiar...

**Comment by mbolivar-nordic** [comment]
> How about this one-liner? Works4me!

Something like this seems reasonable to me. Any issues @aescolar ?

> Way ahead of "us" upstream

We have a lot of tooling that I've been carrying around since my foundries days :slightly_smiling_face: . So far everything has proven to require some domain-specific knowledge, and I haven't bothered to try to untangle reusable from domain-specific. But it's all open source if people want to give it a go and play around.

**Comment by aescolar** [comment]
> Putting a DNM on this until we get the requester to confirm it's what they want, and then I'll write some test cases.

Thank you very much @mbolivar-nordic , that was much faster than I expected.
Overall this solves my problem, so getting it already "as is" is more than welcome.
About skipping the manifest repo, my thinking was that it is still nice to show if it is dirty even if do not have a reference commit. It is conceptually a bit decoupled from the other projects checks, but nonetheless.

Otherwise I only see "extras" that could be added, but those can come later. (say things like providing more concise output than git status, when it is clean but we are in another commit (say a message like "xxxx checked out, manifest-rev was yyyy). Or having the option to get a short (-s) git status)

>> How about this one-liner? Works4me!
> Something like this seems reasonable to me. Any issues @aescolar ?

Ok.

**Comment by marc-hb** [comment]
> About skipping the manifest repo, my thinking was that it is still nice to show if it is dirty even if do not have a reference commit. It is conceptually a bit decoupled from the other projects checks, but nonetheless.

I second this; I would like practically over purity here. It would also be consistent with `west status` which is already a bit "impure" that way.

Or how about this middle-ground: not by default but when some new `west compare --manifest-status` option is passed? Permanently settable in `~/.westconfig` somehow.

**Comment by marc-hb** [comment]
So this additional line shows whether the module is on the `manifest-rev` or not:
```
project.git('log --oneline -n 1 manifest-rev')
```

As a bonus, it also shows whether it's on `upstream/main`  or not.

**Comment by mbolivar-nordic** [comment]
@aescolar @marc-hb you were both totally right about including the manifest repository in the output. I even expected that to be the behavior when I sat down to write the tests tonight, and I was surprised that it wasn't.

Now updated, please give it a go and let me know if it meets your expectations. I will merge after that, and we'll get things rolling for west v1.1 with this feature.

File: src/west/app/project.py [hunk 1]
```diff
@@ -602,6 +602,118 @@ def _dump(self, args, to_dump):
...
+        parser.add_argument('--ignore-branches', action='store_true',
+                            help='''skip output for projects with checked out
+                            branches and clean working trees if the branch is
+                            at the same commit as the last "west update"''')
+        return parser
+
+    def do_run(self, args, ignored):
+        self.die_if_no_git()
+        if self.git_version_info < (2, 22):
+            self.die('git version 2.22 or later is required')
```
Comment (tejlmand) [inline src/west/app/project.py:650]: remember to add to release notes that git >=2.22 is now required if using this command.

Also I believe it would be nice with a comment as to why git >=2.22 is required.
I assume it's to due to the `branch --show-current` which were introduced in 2.22.

But thanks for only having this as requirement for this command 👍 and not enforce it generally for `west`.

File: src/west/app/project.py [hunk 2]
```diff
@@ -602,6 +602,118 @@ def _dump(self, args, to_dump):
...
+               out by the most recent successful "west update"
+            2. its working tree is not clean (i.e. there are local uncommitted
+               changes)
+            3. it has a local checked out branch
+
+            The command prints output for the manifest repository if it
+            has nonempty status.
+
+            The output is meant to be human-readable, and may change. It is
+            not a stable interface to write scripts against.''')
```
Comment (tejlmand) [inline src/west/app/project.py:627]: should we add a line that states that git >=2.22 is required for this command, so that it's also mentioned in the help text. After all, help text is a form of documentation for `west`.
I do know we also have docs here: https://docs.zephyrproject.org/latest/develop/west/index.html, but `west help <cmd>` is faster 😉

**Review by tejlmand — COMMENTED** [review]
Looks good.

Couple of small proposals to be friendlier to users.

**Review by aescolar — APPROVED** [review]
It seems fine :)
(Maybe it would be a bit nicer without the minor banners, or if they did not jump to the next line, but that is quite minor)

**Comment by mbolivar-nordic** [comment]
@tejlmand good points, thanks. I've implemented your suggestions.

**Review by tejlmand — APPROVED** [review]

**Comment by marc-hb** [comment]
I just tested this and it looks pretty good, thanks! Two recommendations:

```diff
modified   src/west/app/project.py
@@ -666,6 +666,7 @@ class Compare(_ProjectCommand):
                         project.git('log --oneline -n 1 HEAD')
                         self.small_banner('status:')
                         project.git('status')
+                        print()
                         printed_status = True
                     except subprocess.CalledProcessError:
                         failed.append(project)
@@ -708,6 +709,7 @@ class Compare(_ProjectCommand):
                 project.git('log --oneline -n 1 manifest-rev')
                 self.small_banner('status:')
                 project.git('status')
+                print()
                 printed_status = True
             except subprocess.CalledProcessError:
                 failed.append(project)
```
... or something like it because `git status` adds this annoying blank line before its last line of output:

```
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	foo

nothing added to commit but untracked files present (use "git add" to track)
=== mbedtls (modules/crypto/mbedtls):
--- manifest-rev:
7fed49c9b9f9 (HEAD, manifest-rev) CC3XX: Hardcode entry points for the CC3XX driver
--- status:
```

**Comment by marc-hb** [comment]
Second recommendation: can you please reverse the `--ignore-branches` logic and change it to `--include-branches` option instead? I really don't see the value of showing a personal branch that is 100% identical to the manifest in a repo that is in a 100% pristine state, this is just adding noise IMHO. The next `west update` command will totally ignore such a branch anyway, so I don't see why `west compare` should show it by default.

While this new `Compare` command looks incredibly useful, its output may get "crowded" and hard to parse pretty quickly (see comments about banners above), so the default should IMHO show only actual code differences that can make a difference and not include any "nice to have" git metadata like "BTW, did you remember you created a git branch there?" when said branch makes no difference whatsoever.

**Comment by aescolar** [comment]
I don't mind the `Second recommendation: can you please reverse the --ignore-branches` request.
But @marc-hb one note about 

> The next west update command will totally ignore such a branch anyway, so I don't see why west compare should show it by default.

Not exactly, the next west update will get out of that branch and detach the head.
 
> While this new Compare command looks incredibly useful, its output may get "crowded" ..
 
Note as mentioned before, I'll be happy to get it already as is. Anyhow with regard to too much verbosity, I guess that is why `repo` has a very concise output: projects that are clean, but on a branch, only produce 1 line output (including the project name, with the branch that is checked out on the same line). And the status per project is a very compact custom notation of the status of each project. But that is minor nicety compared to having the functionality overall, which could be added later.

**Comment by mbolivar-nordic** [comment]
I dithered over this decision for too long and then ended up just adding a chicken bit.

This now supports a compare.ignore-branches configuration option that people can use if they don't like the default behavior, which is still to print information about checked-out branches. (I think surfacing the behavior by default will satisfy people who want it, and people who don't want it will be motivated to figure out how to turn it off)

**Comment by marc-hb** [comment]
> because git status adds this annoying blank line before its last line of output

"fix" submitted in
- #648
